### PR TITLE
feat: Allow opening commit details from status buffer header lines

### PR DESCRIPTION
**Pull request by jarviliam**
This commit allows users to open the commit/ref/tag under their cursor for the top 4 header lines (head,upstream,pushRemote,tags) similar to how magit allows user to do the same on the status buffer. Similarly to how magit displays commits in their viewer, the top line now shows the passed argument (commit,ref,tag) used in the git command. Then below that line is the full commit hash of it.  

Had to modify the section table a bit to allow handling sections where there are no files within but it gets the job done.
Maybe it would be a nicer api in the future to define sections with callbacks for when a user performs a select action on instead of having to check the name of the section and perform an action from there.

Like always feel free to point out changes 🙏 . Also should I be posting in the Neogit discussion/development thread for these kind of changes beforehand?

### CODE REVIEW

**Comment by jarviliam** [comment]
I think the tests are failing because of the previous gpg exec commit? 🤔

**Comment by CKolkey** [comment]
> This commit allows users to open the commit/ref/tag under their cursor for the top 4 header lines (head,upstream,pushRemote,tags) similar to how magit allows user to do the same on the status buffer. Similarly to how magit displays commits in their viewer, the top line now shows the passed argument (commit,ref,tag) used in the git command. Then below that line is the full commit hash of it.
> 

Seriously cool work! Really.

> Had to modify the section table a bit to allow handling sections where there are no files within but it gets the job done. Maybe it would be a nicer api in the future to define sections with callbacks for when a user performs a select action on instead of having to check the name of the section and perform an action from there.
> 

Yeah, the status buffer is... tangled. @ten3roberts is working on some stuff right now that deals with this, so I think she should weigh in before merging this.

> Like always feel free to point out changes 🙏 . Also should I be posting in the Neogit discussion/development thread for these kind of changes beforehand?

Honestly no major changes. Defining the header as a section with no sign drawn is actually a pretty neat way to do this. I'm a bit confused about the failing test.. it's seemingly unrelated, but doesn't fail on master. I'll have a look too.

**Comment by ten3roberts** [comment]
Yes, the status buffer is quite tangled and full of special cases. As you said, currently untangling it bit by bit. The benifit of it is are *less* merge conflicts in that file due to increased containment.

**Comment by ten3roberts** [comment]
> Also should I be posting in the Neogit discussion/development thread for these kind of changes beforehand

If there is a bug your are currently solving it may be worthwhile to open an issue just to make sure nobody else is currently doing the same thing. As for pull requests it may be a good idea if the feature needs discussion of how it should work before work is started :)

File: lua/neogit/status.lua
Comment (ten3roberts) [inline lua/neogit/status.lua:88]: I assume this just makes sure that there is an item (and not a blankline under the cursor?

File: lua/neogit/status.lua
Comment (ten3roberts) [inline lua/neogit/status.lua:1081]: Could this part be moved into a separate function now that it is even larger?

**Review by ten3roberts — CHANGES_REQUESTED** [review]
I think the changes look quite good and as always I really do like the type annotations.

All the changed "selection" logic have clear equivalents in my `feat-ignore` branch which will get merged soon :sweat_smile: 

I got one error when pressing `<ret>` on the `Merge: ` section under your branch. I assume that will be fixed by rebasing on the latest master which includes the fixes for branches containing `/`

On this line:
image

```sh
git --no-pager -c color.ui=always --no-optional-locks show --format=fuller feat:

    fatal: ambiguous argument 'feat': unknown revision or path not in the working tree.
    Use '--' to separate paths from revisions, like this:
    'git <command> [<revision>...] -- [<file>...]'
    

Open the console for details
```

```sh
Error executing vim.schedule lua callback: ...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:18: The coroutine failed with this message: /home/tei/dev/nvim/neogit/lua/neogit/lib/git/log.lua:50: Failed to parse line: 
stack traceback:
	[C]: in function 'error'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:18: in function 'callback_or_next'
	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function <...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:44>
```

**Comment by CKolkey** [comment]
> I think the changes look quite good and as always I really do like the type annotations.
> 
> 
> 
> All the changed "selection" logic have clear equivalents in my `feat-ignore` branch which will get merged soon :sweat_smile: 
> 
> 
> 
> I got one error when pressing `<ret>` on the `Merge: ` section under your branch. I assume that will be fixed by rebasing on the latest master which includes the fixes for branches containing `/`
> 
> 
> 
> On this line:
> 
> image
> 
> 
> 
> ```sh
> 
> git --no-pager -c color.ui=always --no-optional-locks show --format=fuller feat:
> 
> 
> 
>     fatal: ambiguous argument 'feat': unknown revision or path not in the working tree.
> 
>     Use '--' to separate paths from revisions, like this:
> 
>     'git <command> [<revision>...] -- [<file>...]'
> 
>     
> 
> 
> 
> Open the console for details
> 
> ```
> 
> 
> 
> ```sh
> 
> Error executing vim.schedule lua callback: ...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:18: The coroutine failed with this message: /home/tei/dev/nvim/neogit/lua/neogit/lib/git/log.lua:50: Failed to parse line: 
> 
> stack traceback:
> 
> 	[C]: in function 'error'
> 
> 	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:18: in function 'callback_or_next'
> 
> 	...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:45: in function <...share/nvim/lazy/plenary.nvim/lua/plenary/async/async.lua:44>
> 
> ```

Yea, i noticed that too and just merged a fix on master since its just a general bug. A rebase will solve that (I think) :)

File: lua/neogit/status.lua
Comment (jarviliam) [reply lua/neogit/status.lua:1081]: 👍

**Comment by jarviliam** [comment]
Turns out the failing test was because the test was calling `Ma<CR>` and after a the buffer goes back to the status page. Then the <CR> input toggles the head commit view. Removing the <CR> input fixes it 👍

**Review by CKolkey — APPROVED** [review]

**Comment by CKolkey** [comment]
Looks good to me. Any objections @ten3roberts ?

**Review by ten3roberts — APPROVED** [review]
Looks amazing :smile:

**Pull request merged**
